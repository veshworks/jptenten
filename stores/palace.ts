import { createPalace } from '@any-percent/engine';
import { get } from '@vhoyer/utils/dist/object';
import { diff } from 'deep-diff';
import Tormenta20 from '~/assets/bundles/tormenta-20/index.js?raw';

const palace = createPalace([Tormenta20]);

type FormLocator = { entityId: string; index: number; };

export const usePalaceStore = defineStore('palace', () => {
  const answers = ref({} as Record<string, any[]>);

  function setAnswer({ entityId, index }: FormLocator, value: any) {
    answers.value[entityId] ||= [];
    answers.value[entityId][index] = value;
  }

  const builder = ref({
    sheet: {},
    form: {},
  } as ReturnType<typeof palace.builder.createSheet>);

  const newBuilder = ref({
    sheet: {},
    form: {},
  } as ReturnType<typeof palace.builder.createSheet>);

  watch(answers, () => {
    newBuilder.value = palace.builder.createSheet(answers.value);
  }, { deep: true });

  const builderDiff = computed(() => diff(builder.value, newBuilder.value) ?? []);

  watch(builderDiff, () => {
    for (const diffItem of builderDiff.value ?? []) {
      const path: string[] = diffItem.path ?? [];

      let NECK: any = null;
      let HEAD: any = builder.value;
      const headName: string = path.at(-1) ?? '';

      // select the HEAD of the object
      for (const part of path) {
        NECK = HEAD;
        HEAD = HEAD[part];
      }

      // get the new value
      const newValue = get(newBuilder.value, path);
      // apply to reactive object
      NECK[headName] = newValue;
    }
  }, { immediate: true });

  const sheet = computed(() => builder.value.sheet);
  const form = computed(() => builder.value.form);

  return {
    // private
    _answers: answers,
    _builder: builder,
    _newBuilder: newBuilder,
    _builderDiff: builderDiff,

    // public
    sheet,
    form,
    setAnswer,
    getEntityById: (idRaw: MaybeRefOrGetter<string>) => {
      const id = toValue(idRaw)
        .replace(/^\//g, '')
        .replace(/\/$/g, '')
        .split('/')
        .at(-1) ?? '';
      return palace.getEntityById(id);
    },
  };
});
