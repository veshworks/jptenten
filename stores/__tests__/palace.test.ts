// @vitest-environment nuxt
import { setActivePinia, createPinia } from 'pinia';
import { describe, it, expect, beforeEach } from 'vitest';
import { usePalaceStore } from '../palace';

describe('Store > palace', () => {
  beforeEach(() => {
    setActivePinia(createPinia());
  });

  it('exposes public API correctly', () => {
    const palace = usePalaceStore();

    expect(palace.sheet).toMatchSnapshot();
    expect(palace.form).toMatchSnapshot();
    expect(palace.setAnswer).toBeTypeOf('function');
    expect(palace.getEntityById).toBeTypeOf('function');
  });

  describe('#getEntityById', () => {
    it('returns the entity', () => {
      const palace = usePalaceStore();
      expect(palace.getEntityById('attribute.for')).toMatchSnapshot();
    });
  });

  describe('setAnswer method=manual', () => {
    beforeEach(() => {
      const palace = usePalaceStore();
      palace.setAnswer({ entityId: '/aspect.attributes/', index: 0 }, 'manual');
    });

    it('changes the value of form', () => {
      const palace = usePalaceStore();
      expect(palace.form).toMatchSnapshot();
    });

    describe('setAnswer attributes=obj', () => {
      beforeEach(() => {
        const palace = usePalaceStore();
        const attributeObject = JSON.parse(JSON.stringify(
          palace.form['/aspect.attributes/'].fields[1].attributes,
        )) as Record<string, any>;

        // assign index to each attribute
        Object.keys(attributeObject).forEach((key, index) => {
          attributeObject[key] = index;
        });

        palace.setAnswer({ entityId: '/aspect.attributes/', index: 1 }, attributeObject);
      });

      it('changes the value of sheet', () => {
        const palace = usePalaceStore();
        expect(palace.sheet).toMatchSnapshot();
      });
    });
  });
});
