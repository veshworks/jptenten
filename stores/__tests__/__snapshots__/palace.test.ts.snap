// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Store > palace > #getEntityById > returns the entity 1`] = `
{
  "builderBehavior": undefined,
  "composePhase": false,
  "meta": {
    "description": "Seu poder muscular. A Força é aplicada em testes de Atletismo e Luta; rolagens de dano corpo a corpo ou com armas de arremesso, e testes de Força para levantar peso e atos similares.",
    "pools": [
      "attribute",
    ],
    "title": "Força (FOR)",
  },
}
`;

exports[`Store > palace > exposes public API correctly 1`] = `
{
  "attributes": {
    "car": 0,
    "con": 0,
    "des": 0,
    "for": 0,
    "int": 0,
    "sab": 0,
  },
}
`;

exports[`Store > palace > exposes public API correctly 2`] = `
{
  "/aspect.attributes/": {
    "fields": [
      {
        "default": "points",
        "description": "Você pode definir seus atributos de três formas diferentes.

**Pontos:** Você começa com todos os atributos em 0 e recebe 10 pontos para comprar ou vender os atributos.

**Rolando:** Rolamos automaticamente os valores e você pode distribuir os valores nos atributos que quiser.

**Manual:** Preencha manualmente cada um dos campos.",
        "label": "Método de seleção",
        "options": [
          {
            "display": "Gastando pontos",
            "value": "points",
          },
          {
            "display": "Rolando dados",
            "value": "roll",
          },
          {
            "display": "Insersão Manual",
            "value": "manual",
          },
        ],
        "type": "combo",
      },
      {
        "attributes": {
          "attribute.car": "Carisma",
          "attribute.con": "Constituição",
          "attribute.des": "Destreza",
          "attribute.for": "Força",
          "attribute.int": "Inteligência",
          "attribute.sab": "Sabedoria",
        },
        "label": "Definindo Seus Atributos",
        "method": "points",
        "points": {
          "cost": {
            "-1": -1,
            "0": 0,
            "1": 1,
            "2": 2,
            "3": 4,
            "4": 7,
          },
          "label": "pontos",
          "maxValue": 4,
          "minValue": -1,
          "starting": 10,
        },
        "rolling": "{(4d6dl1-10)/2, {-2}}kh1",
        "type": "attributes",
      },
    ],
    "status": "warn",
  },
}
`;

exports[`Store > palace > setAnswer method=manual > changes the value of form 1`] = `
{
  "/aspect.attributes/": {
    "fields": [
      {
        "default": "points",
        "description": "Você pode definir seus atributos de três formas diferentes.

**Pontos:** Você começa com todos os atributos em 0 e recebe 10 pontos para comprar ou vender os atributos.

**Rolando:** Rolamos automaticamente os valores e você pode distribuir os valores nos atributos que quiser.

**Manual:** Preencha manualmente cada um dos campos.",
        "label": "Método de seleção",
        "options": [
          {
            "display": "Gastando pontos",
            "value": "points",
          },
          {
            "display": "Rolando dados",
            "value": "roll",
          },
          {
            "display": "Insersão Manual",
            "value": "manual",
          },
        ],
        "type": "combo",
      },
      {
        "attributes": {
          "attribute.car": "Carisma",
          "attribute.con": "Constituição",
          "attribute.des": "Destreza",
          "attribute.for": "Força",
          "attribute.int": "Inteligência",
          "attribute.sab": "Sabedoria",
        },
        "label": "Definindo Seus Atributos",
        "method": "manual",
        "points": {
          "cost": {
            "-1": -1,
            "0": 0,
            "1": 1,
            "2": 2,
            "3": 4,
            "4": 7,
          },
          "label": "pontos",
          "maxValue": 4,
          "minValue": -1,
          "starting": 10,
        },
        "rolling": "{(4d6dl1-10)/2, {-2}}kh1",
        "type": "attributes",
      },
    ],
    "status": "warn",
  },
}
`;

exports[`Store > palace > setAnswer method=manual > setAnswer attributes=obj > changes the value of sheet 1`] = `
{
  "attributes": {
    "car": 5,
    "con": 2,
    "des": 1,
    "for": 0,
    "int": 3,
    "sab": 4,
  },
}
`;
